tpie/pipelining/node.cpp

```diff
@@ -170,30 +163,31 @@ void node::add_dependency(const node & dest) {
 	add_dependency(dest.token);
 }
 
-void node::set_minimum_memory(memory_size_type minimumMemory) {
-	switch (get_state()) {
-	case STATE_IN_PROPAGATE:
-	case STATE_AFTER_PROPAGATE:
-	case STATE_FRESH:
-	case STATE_IN_PREPARE:
-		break;
-	default:
-		throw call_order_exception("set_minimum_memory");
-	}
-	m_parameters.minimumMemory = minimumMemory;
-}
 
-void node::set_maximum_memory(memory_size_type maximumMemory) {
-	switch (get_state()) {
-	case STATE_IN_PROPAGATE:
-	case STATE_AFTER_PROPAGATE:
-	case STATE_FRESH:
-	case STATE_IN_PREPARE:
-		break;
-	default:
-		throw call_order_exception("set_maximum_memory");
+#define PARAM_SETTER(setter_name, param_type, param_name) \
+	void node::setter_name(param_type param_name) { \
+		switch (get_state()) { \
+		case STATE_IN_PROPAGATE: \
+		case STATE_AFTER_PROPAGATE: \
+		case STATE_FRESH: \
+		case STATE_IN_PREPARE: \
+			break; \
+		default: \
+			throw call_order_exception(#setter_name); \
+		} \
+		m_parameters.param_name = param_name; \
 	}
-	m_parameters.maximumMemory = maximumMemory;
+
+PARAM_SETTER(set_minimum_file_descriptors, int, minimumFileDescriptors);
+PARAM_SETTER(set_maximum_file_descriptors, int, maximumFileDescriptors);
+PARAM_SETTER(set_file_descriptor_fraction, double, fileDescriptorFraction);
+
+PARAM_SETTER(set_minimum_memory, memory_size_type, minimumMemory)
+PARAM_SETTER(set_maximum_memory, memory_size_type, maximumMemory)
+PARAM_SETTER(set_memory_fraction, double, memoryFraction);
+
+void node::set_available_file_descriptors(int availableFileDescriptors) {
+	m_availableFileDescriptors = availableFileDescriptors;
 }
```

Comment:
Macros should start with TPIE_ to not conflict with other headers, also it would be nice to undef the macro once we are done with it.  Perhaps instead of a macro we could use a template? Like

template <typename T>
void setter(T & parm, T v) {
	switch (get_state()) { 
	case STATE_IN_PROPAGATE: 
	case STATE_AFTER_PROPAGATE: 
	case STATE_FRESH: 
	case STATE_IN_PREPARE: 
	break; 
	default: 
		throw call_order_exception("");
	} 
	parm = v;
}

void set_minimum_file_destriptions(int v) {setter(minimumFileDescriptions, v);}